public/app/features/explore/state/reducers.ts

```diff
@@ -367,15 +367,24 @@ export const itemReducer = (state: ExploreItemState = makeExploreItemState(), ac
   }
 
   if (removeQueryRowAction.match(action)) {
-    const { queries, queryKeys } = state;
+    const { queries } = state;
     const { index } = action.payload;
 
     if (queries.length <= 1) {
       return state;
     }
 
-    const nextQueries = [...queries.slice(0, index), ...queries.slice(index + 1)];
-    const nextQueryKeys = [...queryKeys.slice(0, index), ...queryKeys.slice(index + 1)];
+    // removes a query under a given index and reassigns query keys and refIds to keep everything in order
+    const queriesAfterRemoval: DataQuery[] = [...queries.slice(0, index), ...queries.slice(index + 1)].map(query => {
+      return { ...query, refId: '' };
+    });
+    const nextQueries: DataQuery[] = [];
+
+    queriesAfterRemoval.forEach((query, i) => {
+      nextQueries.push(generateNewKeyAndAddRefIdIfMissing(query, nextQueries, i));
+    });
```

Comment:
You could use `.map()` instead of `forEach()` and `push()` here and skip the empty array initialization.

Reply:
@s-h-a-d-o-w sure, but we'll need to write yet another helper or significantly modify the existing ones not to use partial results (hence the array) to assign refIds properly. No performance difference though - we're talking about queries array and its length usually doesn't exceed 30.

Reply:
Sorry, I missed something about how this works when I read it. Forget about my comment. :)